### PR TITLE
Trackers Bucketing (based on `sensitive_pixel_options` cookie)

### PR DESCRIPTION
**Pull request by robertsreberski**
- Implement bucketing (based on a8c-analytics.js)
- Add support for `region` cookie

#### Proposed Changes

* Implement recognition of trackers bucketing cookie (`essential`, `advertising`, `analytics`)
* Implement code that parses `sensitive_pixel_options` to determine if the specific bucket can be tracked
* Replace simple enabled/disabled constant checks for ad trackers with `tracker-buckets.ts` check functions (i.e. `mayWeTrackByTracker`)

#### Testing Instructions

* Ensure `ad-tracking` is enabled on your development environment
* Launch Calypso / Jetpack Cloud
* If not based in CCPA/GDPR region, set a cookie e.g. `document.cookie = 'country_code=NL'` (or any other European country)
* Make sure you have you don't have "sensitive_pixel_option" (legacy) cookie in your session: `document.cookie='sensitive_pixel_option=; expires=Thu, 01 Jan 1970 00:00:01 GMT;'`
* Manually create a bucketing cookie (all buckets accepted): `document.cookie = 'sensitive_pixel_options=%7B%22version%22%3A20201224%2C%22ok%22%3Atrue%2C%22buckets%22%3A%7B%22essential%22%3Atrue%2C%22analytics%22%3Atrue%2C%22advertising%22%3Atrue%7D%7D';`
* Notice that GA is being loaded and works as usual i.e. if `https://www.googletagmanager.com/gtag/js?id=` scripts are included in page DOM)
* Manually change a bucketing cookie to exclude `analytics` bucket: `document.cookie = 'sensitive_pixel_options=%7B%22version%22%3A20201224%2C%22ok%22%3Atrue%2C%22buckets%22%3A%7B%22essential%22%3Atrue%2C%22analytics%22%3Afalse%2C%22advertising%22%3Atrue%7D%7D';`
* Notice that GA stopped being loaded to the session, and the rest of trackers (e.g. FB Pixel) is still working i.e. if `https://connect.facebook.net/signals/config/` script is included in page DOM
* Manually change a bucketing cookie to exclude `advertising` bucket: `document.cookie = 'sensitive_pixel_options=%7B%22version%22%3A20201224%2C%22ok%22%3Atrue%2C%22buckets%22%3A%7B%22essential%22%3Atrue%2C%22analytics%22%3Atrue%2C%22advertising%22%3Afalse%7D%7D';`
* Notice that GA works again, but the rest of trackers (i.e. Facebook Pixel) does not load anymore

To Do:
- [x] Decide if Jetpack GA does use remarketing or not (p1666809890242429-slack-C01264051NE) and update the tracker bucket accordingly

Related: pbNhbs-3sm-p2

### CODE REVIEW

File: client/lib/analytics/utils/get-tracking-prefs.ts
Comment (gmovr) [inline client/lib/analytics/utils/get-tracking-prefs.ts:39]: Here, we can also assume users in California (and other relevant states) are ok with all cookies, until they opt out. Only in GDPR are these pixels opt-in, while in the new US laws, these are still opt-out.

We can assume all buckets are true, until the user has opted out (also for CA, UT, CO, CT, etc.)

File: client/lib/analytics/utils/is-region-in-ccpa-zone.ts
Comment (gmovr) [inline client/lib/analytics/utils/is-region-in-ccpa-zone.ts:2]: I like this logic, makes it easy to add new states (a lot more states will adapt some form of privacy laws). California is the strictest, so if we treat any new state as CA, we'll be good.

**Review by gmovr — COMMENTED** [review]
Left some comments. This approach looks good to me, and it's very flexible and extendable, i.e. as laws change. 

Will do some more manual testing.

File: client/lib/analytics/ad-tracking/setup.js
Comment (CodeyGuyDylan) [inline client/lib/analytics/ad-tracking/setup.js:36]: When I have the cookie that disables google analytics: `document.cookie = 'sensitive_pixel_options=%7B%22version%22%3A20201224%2C%22ok%22%3Atrue%2C%22buckets%22%3A%7B%22essential%22%3Atrue%2C%22analytics%22%3Afalse%2C%22advertising%22%3Atrue%7D%7D';`
I am getting an error thrown here. It seems like `mayWeTrackByTracker( 'googleAds' )` is still returning true even though the cookie has analytics disabled

Comment (CodeyGuyDylan) [reply client/lib/analytics/ad-tracking/setup.js:36]: Here is the error:
image

**Review by CodeyGuyDylan — CHANGES_REQUESTED** [review]
The first and third cookie example tested correctly, however, I got an error thrown on one of them, I made the comment where the error was thrown

File: client/lib/analytics/ad-tracking/setup.js
Comment (CodeyGuyDylan) [reply client/lib/analytics/ad-tracking/setup.js:36]: FWIW, this does completely stop the page from being loaded for me. I have country_code set to 'US' and region set to 'california'

Comment (robertsreberski) [reply client/lib/analytics/ad-tracking/setup.js:36]: Thanks, @CodeyGuyDylan for spotting that one out, it appears the floodlight and google ads setup were incomplete. It should be fixed with the following commit!

File: client/lib/analytics/tracker-buckets.ts
Comment (jboland88) [inline client/lib/analytics/tracker-buckets.ts]: Would it be helpful to use `Bucket.ADVERTISING` here to take advantage of the string enum you have set up for `Bucket`?

File: client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js
Comment (jboland88) [inline client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js]: Nitpick - since you deal also with the `region` cookie here now, is it helpful to show that value in the debug message too?

File: client/lib/analytics/utils/get-tracking-prefs.ts
Comment (jboland88) [inline client/lib/analytics/utils/get-tracking-prefs.ts:39]: Does the logic order matter here? Noticing that the previous `mayWeTrackCurrentUserGdpr ` function checked the value of `sensitive_pixel_option` prior to checking the `countryCode`.  Not sure if it's possible that both `isCountryInGdprZone` and `isRegionInCcpaZone` return false and that there is a cookie value set for tracker preferences. Would it make sense to check the preferences cookies first?

File: client/lib/analytics/tracker-buckets.ts
Comment (jboland88) [inline client/lib/analytics/tracker-buckets.ts]: Are these TODO items complete? I see some code related to GAds for different environments.

File: client/lib/analytics/utils/get-tracking-prefs.ts
Comment (jboland88) [inline client/lib/analytics/utils/get-tracking-prefs.ts:64]: Would it be better to set `buckets` to `allBucketsFalse` if `oldUserConsent` is `'no'`?

File: client/lib/analytics/ad-tracking/load-tracking-scripts.js
Comment (jboland88) [inline client/lib/analytics/ad-tracking/load-tracking-scripts.js]: Is `iconMedia` the right one to check? It looks like the bing script is loaded here?

File: client/lib/analytics/ad-tracking/setup.js
Comment (jboland88) [inline client/lib/analytics/ad-tracking/setup.js:202]: Is `mayWeTrackByTracker` also needed here?

**Review by jboland88 — CHANGES_REQUESTED** [review]
I was not fully sure how to test and verify that different tracking was/ was not loaded. 

For steps like this one in the test plan:
> Notice that GA is being loaded and works as usual

Could you provide some more instruction on what to check to verify that the script is loaded and working? I was able to check the source in some instances to see if the appropriate `<script />` was present, but there may be a better way to verify?

Another question - If I am testing in a non CCPA region in the US, what should the behavior be? In this scenario in my testing, I don't think any tracking scripts loaded, but I was not sure how to verify this for sure.

Overall the code looks pretty good - had a few questions and nitpicks that I left inline.

@gmovr - if you have time to review again after changes, you may see things that I did not :)

File: client/lib/analytics/ad-tracking/setup.js
Comment (robertsreberski) [reply client/lib/analytics/ad-tracking/setup.js:202]: `setupWpcomGoogleAdsGtag()` and `setupWpcomFloodlightGtag()` are functions invoked in the code defined above. It already includes proper checks before invoking them (see lines 36 and 40).

File: client/lib/analytics/utils/get-tracking-prefs.ts
Comment (robertsreberski) [reply client/lib/analytics/utils/get-tracking-prefs.ts:64]: The old user consent was binary - either you accepted all or none cookies. If `ok === true`, we check specific buckets - and in this case, we want all of them to be `true` to match the old way of user consent. if `ok === false`, bucket values don't matter - we don't check them.

I don't have a preference towards either setting all buckets to `true` or to `false`. However, I believe `allBucketsTrue` is as much accurate as `allBucketsFalse`.

File: client/lib/analytics/utils/get-tracking-prefs.ts
Comment (robertsreberski) [reply client/lib/analytics/utils/get-tracking-prefs.ts:39]: Thanks for pointing out that change @jboland88. We might ask @gmovr for an opinion, but geofencing is more important than the current cookie value. So if the user moves to a country where GDPR / CCPA does not apply - we do not want to comply with the cookie setting anymore.

Note that the code will not overwrite the cookie, so if the user moves back to the compliant country/state, and get through the geofencing check - other cookie conditions will apply.

File: client/lib/analytics/ad-tracking/load-tracking-scripts.js
Comment (robertsreberski) [reply client/lib/analytics/ad-tracking/load-tracking-scripts.js]: Good catch, thanks!

File: client/lib/analytics/tracker-buckets.ts
Comment (robertsreberski) [reply client/lib/analytics/tracker-buckets.ts]: Thanks for pointing that out - indeed these TODOs are complete - they were about ensuring that GAds for both wpcom and jetpack environments are handled correctly - and we [enable them](https://github.com/Automattic/wp-calypso/pull/66493/files#diff-a520458bb03bb680ec8f6c94e18093080bd8524132146eb463504977f12654d6L12-L13) in both environments. See [this function](https://github.com/Automattic/wp-calypso/pull/66493/files#diff-1b55ef351c80a1afa5c4533f2da21e31c038bd1525cfca17d809c1d7de1fbcacR384) for reference.

**Comment by robertsreberski** [comment]
Thanks for your insight into the code @jboland88! Indeed, we'll not merge it until @gmovr will also approve changes - given the experience in that area - however, I believe reviews from our team are very insightful as well!

> Could you provide some more instruction on what to check to verify that the script is loaded and working? I was able to check the source in some instances to see if the appropriate <script /> was present, but there may be a better way to verify?

Thanks for your feedback on the testing plan. Indeed, I assumed that it's clear; I'm correcting it so as to be more explicit.

> Another question - If I am testing in a non CCPA region in the US, what should the behavior be? In this scenario in my testing, I don't think any tracking scripts loaded, but I was not sure how to verify this for sure.

Then the 3rd point of the testing plan applies (setting cookie country is one example solution to that issue, but it's more bulletproof for testers worldwide). Or do you think we need to have some debug value here?

**Review by gmovr — COMMENTED** [review]
I tested via the following URL pattern: https://container.calypso.live/read?flags=google-analytics,ad-tracking and set my country code to NL. 

I noticed that I still have to accept the cookie banner (or manually set `sensitive_pixel_option=yes` for it to work. 

I went through the examples, and they all work. 

For WPcom Google Analytics is under "advertising" due to (potentially) remarketing. So, by excluding advertising trackers, GA was also disabled. 

But wondering why this is, as it appears `mayWeTrackByTracker( 'ga' )` should fire as long as analytics is set to true. E.g. in production, if I set the analytics and essential buckets to true, along with the old `sensitive_pixel_option` consent to "yes" I can see pageviews on GA4 tracking.

**Comment by robertsreberski** [comment]
Thanks for the review @gmovr!

You shouldn't need to set the `sensitive_pixel_option=yes` legacy cookie, which overwrites reading the new version of privacy preferences (`sensitive_pixel_options`). I also clarified a testing plan by including a step to ensure this legacy cookie does not exist in your session.

> So, by excluding advertising trackers, GA was also disabled.

For specific actions, we were stopping the execution of the function when advertising trackers were disabled (so for some pages / actions, it was not reaching the point of loading the ga script). What is wrong, we can depend on the checks of the specific trackers - that way we'll avoid situations to categorize analytics cookie as advertising implicitly. 

I've also changed the bucket category for GA - now it's `advertising` for WPCOM, and `analytics` for JP Cloud - as defined in our Cookie Jar.

**Review by gmovr — APPROVED** [review]
Thanks! I did another round of testing, and all works well for me!

I also tested as a non-GDPR user (Canada and US), and it's working well (all trackers fire by default, which is what we want). 

Testing in GDPR, all the steps worked (e.g. disabling analytics/advertising, or allowing all). 

I also tested some logged out routes, e.g. /start/ and all is working as expected there as well. 

🚢

**Comment by gmovr** [comment]
This test suite is failing `/lib/analytics/` I can take a look at it too. :)

**Review by CodeyGuyDylan — APPROVED** [review]
Everything looks good to me! I think this is a much better way that it was done before by the looks of it

File: client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js
Comment (jboland88) [inline client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js]: Any reason why `country_code` is snake case in the error return and camel case in the success return?

File: client/lib/analytics/utils/get-tracking-prefs.ts
Comment (jboland88) [reply client/lib/analytics/utils/get-tracking-prefs.ts:39]: > geofencing is more important than the current cookie value. So if the user moves to a country where GDPR / CCPA does not apply - we do not want to comply with the cookie setting anymore.

Gotcha, thanks for the info there!

File: client/lib/analytics/utils/get-tracking-prefs.ts
Comment (jboland88) [reply client/lib/analytics/utils/get-tracking-prefs.ts:64]: > if ok === false, bucket values don't matter - we don't check them.

Makes sense, probably fine to leave as is then.

My suggestion was mainly a cautious one in case, for some reason, a future contributer decided to check the `buckets` value without first checking the `ok` value.

**Review by jboland88 — COMMENTED** [review]
Following up on existing comments and noted one more small question. I have not re-run the functional testing myself.

File: client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js
Comment (robertsreberski) [reply client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js]: Thanks for spotting that typo!

**Comment by sirbrillig** [comment]
FYI I'm still not sure how from reading the code, but this PR appears to have caused a bug (or maybe just exposed an existing issue) where some calls to `window.gtag()` failed because `window.gtag` was not defined. I discovered this because we have extra reporting for fatal calypso errors that happen during checkout, but these errors could be happening in other places within calypso as well.

I merged the following PRs to fix the issue for checkout:

- https://github.com/Automattic/wp-calypso/pull/70182
- https://github.com/Automattic/wp-calypso/pull/70185

And I prepared this follow-up as well, although the fatals we are seeing appear to have been stopped by the above changes https://github.com/Automattic/wp-calypso/pull/70187

**Comment by robertsreberski** [comment]
@sirbrillig Thanks for dealing with the issue! I agree that cookie preferences are not enough to guard this code (as it also depends on browser / ad blockers active etc.) and will improve the code based on that. I'll create a follow-up PR with further fixes.

**Pull request merged**
